### DISCUSSION TITLE
binary packages support

### DISCUSSION THREAD

**Discussion by rkitover** [microsoft/vcpkg#18332] [Ideas]
I wrote a prototype for supporting binary packages in vcpkg in PowerShell:

https://github.com/rkitover/vcpkg-binpkg-prototype/

What do you guys think? Would it be worthwhile to work on this in the core?

**Comment by BillyONeal** [comment]
<s>Packaging binaries is somewhat incompatible with our design goal of ensuring that the tree remains functional when any given port is updated because we can rebuild everything within that port's "cone of destruction". These kinds of issues are the primary reason why "NuGet for C++" became defunct and why we built vcpkg in the first place.</s>

On second thought it seems like this might not be an attempt to stably package binaries? I apologize that I haven't followed exactly what your other repo is trying to do that closely...

**Comment by BillyONeal** [comment]
Does the official binary caching feature @ras0219 and @ras0219-msft submitted described in https://github.com/microsoft/vcpkg/pull/11204 implement what you wanted here?

**Comment by rkitover** [comment]
My module actually makes perfectly functional binary packages, I need to make a few more tweaks to add the `Abi:` key etc., I'm actually thinking of making a binary package server.

**Comment by JackBoosY** [comment]
I think we will never provide pre-compiled binary packages.

**Comment by rkitover** [comment]
I'm actually considering switching to Conan for a project because it does provide binary packages and things like recompiling ffmpeg can get very tedious, plus it causes a lot of issues with the CI.

**Comment by JackBoosY** [comment]
@rkitover  Yep, I know that. but providing binary packages represents a lack of code security.

**Comment by BillyONeal** [comment]
I think we might provide binaries someday, but if we do it would be through the existing binary caching feature.

(It is a shame that providing binaries is so licensing-scary)

**Reply by autoantwort** [reply]
Yeah would be cool if you can make your existing binary cache freely accessible :)

**Reply by BillyONeal** [reply]
If you want to put the lawyers children through law school....

**Reply by Be-ing** [reply]
This could be easier if vcpkg required SPDX license identifiers in new/updated ports.

**Comment by LegalizeAdulthood** [comment]
I am interested in this because there are some libraries that are freely available and can be used freely by non-commercial applications, but aren't available in source code form.  It would be nice if vcpkg could support these as it essentially eliminates the need to write home brew CMake Find modules and having to specify locations to CMake when configuring.

One example is the [BASS audio library](https://www.un4seen.com/).

At the very least I think it would be instructive to document how to support such libraries within vcpkg so that 3rd party port repositories can support these sorts of libraries.

**Reply by Osyotr** [reply]
> At the very least I think it would be instructive to document how to support such libraries within vcpkg so that 3rd party port repositories can support these sorts of libraries.

You can write portfile.cmake in such way that it only downloads prebuilt library, unpacks it and puts headers, libs and DLLs into corresponding folders.
Be careful though, because you will need to make sure your library is abi compatible with your code.

**Reply by LegalizeAdulthood** [reply]
> You can write portfile.cmake in such way that it only downloads prebuilt library

Is there any documentation or an example on how to do this?

Thanks

**Reply by BillyONeal** [reply]
> > You can write portfile.cmake in such way that it only downloads prebuilt library
> 
> Is there any documentation or an example on how to do this?
> 
> Thanks

There are examples, like `chartdir`: it just downloads the package, unzips it, then `file(COPY`s the right bits into place. I don't know if we have explicit documentation of "here's how you do that" because we discourage ports doing that as it makes the catalog brittle. (As in, this is why NuGet for C++ is a defunct project)
